ngraph/core/src/op/reshape.cpp

```diff
@@ -54,6 +54,122 @@ namespace
         }
     }
+
+    void calculate_output_shape(const ngraph::op::v1::Reshape* reshape_node,
+                                vector<Dimension>& reshape_pattern,
+                                const int64_t& minus_one_idx,
+                                const PartialShape& input_pshape,
+                                vector<Dimension>& output_shape)
+    {
+        if (reshape_pattern == std::vector<Dimension>{0} && !reshape_node->get_special_zero())
+        { // legacy check introduced by PR #1206
+            reshape_pattern = std::vector<Dimension>{};
+            output_shape = {};
+            return;
+        }
+        Dimension output_product(1);
+        for (size_t i = 0; i < reshape_pattern.size(); ++i)
+        {
+            if (i == minus_one_idx) // resolving everything except -1
+                continue;
+
+            auto pattern_dim = reshape_pattern[i];
+            if (pattern_dim.get_min_length() == 0 && pattern_dim.get_max_length() == 0 &&
+                reshape_node->get_special_zero())
+            {
+                if (input_pshape.rank().is_dynamic())
+                {
+                    output_shape[i] = Dimension::dynamic();
+                    output_product *= Dimension::dynamic();
+                }
+                else
+                {
+                    NODE_VALIDATION_CHECK(reshape_node,
+                                          i < input_pshape.rank().get_length(),
+                                          "'0' dimension is out of range");
+                    output_shape[i] = input_pshape[i];
+                    // we do not include dimension to output product here and won't include in input
+                    // product later because we will divide output_product by input_product. This
+                    // dimension contributes to both products equally, but in case this dimension
+                    // is dynamic and others are not we could fully define output dimension that
+                    // is masked by -1
+                }
+            }
+            else
+            {
+                // Dimension(Interval) is a private ctor, so we use public one here
+                output_shape[i] = pattern_dim;
+                output_product *= pattern_dim;
+            }
+        }
+        Dimension input_product(1);
+        if (input_pshape.rank().is_static())
+            for (size_t i = 0; i < input_pshape.rank().get_length(); ++i)
+            {
+                if (i < reshape_pattern.size() && reshape_pattern[i] == 0)
+                    continue;
+                input_product *= input_pshape[i];
+            }
+        else
+            input_product = Dimension::dynamic();
+
+        if (minus_one_idx != -1) // resolving -1 masked dimension
+        {
+            if (output_product.get_min_length() == 0 && output_product.get_max_length() == 0)
+            {
+                // TODO(amprocte): Decide if this is desired behavior here. (NumPy seems
```

Comment:
Do we still need this comment? 😄

Reply:
I don't like the idea of losing this information, cause I wasn't resolving this comment in current review